{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <i>Fusarium oxysporum</i> Inulinase Visualisation With Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [**Biopython Tutorial and Cookbook**](http://biopython.org/DIST/docs/tutorial/Tutorial.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Table of contents:<br>\n",
    "1. [Attributes of Biopython](#1.-Attributes-of-Biopython)\n",
    "2. [Understand FASTA file format](#2.-Understanding-FASTA-file-format)\n",
    "3. [Sequence manipulation using Biopython](#3.-Sequence-manipulation-using-Biopython)\n",
    "4. [Transcription & Translation](#4.-Transcription-&-Translation)\n",
    "5. [Perform Basic Local Alignment using NCBI-BLAST](#5.-Basic-Local-Alignment-Using-NCBI-BLAST)\n",
    "6. [Reading PDB file](#6.-Reading-PDB-file)\n",
    "7. [<i>Fusarium oxysporum</i> Inulunase Protein structure](#7.-<i>Fusarium-oxysporum</i>-Inulunase-Protein-structure)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a3c5cdd4f69848fe96db3e721f6e9851",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import Bio\n",
    "import pylab\n",
    "import urllib\n",
    "import pandas as pd\n",
    "import nglview as nv\n",
    "import numpy as np\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import NCBIWWW\n",
    "from collections import Counter\n",
    "from Bio.Data import CodonTable\n",
    "from Bio import SeqIO, SearchIO\n",
    "from Bio.PDB import PDBParser,MMCIFParser\n",
    "from Bio.SeqUtils import GC, molecular_weight"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Attributes of Biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Align',\n",
       " 'BiopythonDeprecationWarning',\n",
       " 'BiopythonExperimentalWarning',\n",
       " 'BiopythonParserWarning',\n",
       " 'BiopythonWarning',\n",
       " 'Blast',\n",
       " 'Data',\n",
       " 'File',\n",
       " 'GenBank',\n",
       " 'MissingExternalDependencyError',\n",
       " 'MissingPythonDependencyError',\n",
       " 'PDB',\n",
       " 'SVDSuperimposer',\n",
       " 'SearchIO',\n",
       " 'Seq',\n",
       " 'SeqFeature',\n",
       " 'SeqIO',\n",
       " 'SeqRecord',\n",
       " 'SeqUtils',\n",
       " 'Sequencing',\n",
       " 'StreamModeError',\n",
       " 'SwissProt',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " '__version__',\n",
       " '_parent_dir',\n",
       " 'os',\n",
       " 'warnings']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check Attributes of Biopython\n",
    "dir(Bio)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Understanding FASTA file format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GENETIC MATERIAL\n",
    "\n",
    "<div class=\"alert alert-block alert-info\" style=\"Font-size:16px\">\n",
    "\n",
    "**DNA** includes **A T G C**<br>\n",
    "**RNA** includes **U** instead of **T**<br>\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1. FASTA Format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTA File\n",
    "\n",
    "    \n",
    "**Description_of_DNA_sequence**\n",
    "<br>\n",
    "ATGCTGCGAGACAGACAGACATACTATCATCTCAGACGCAG<br>\n",
    "ATGCTGCGAGACAGACAGACATACTATCATCTCAGACGCAG<br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2. Reading from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('ATGATTACATATCTCAGTCACAAGGCCTGGTTCATGCCTTATCCTACGAAGTTC...TGA'), id='NW_017264853.1:16178-17746', name='NW_017264853.1:16178-17746', description='NW_017264853.1:16178-17746 FOXG_17513 [organism=Fusarium oxysporum f. sp. lycopersici 4287] [GeneID=28958268] [chromosome=Unknown]', dbxrefs=[])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inulinase = SeqIO.read(\"gene.fna\", \"fasta\")\n",
    "inulinase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Bio.SeqRecord.SeqRecord"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(inulinase)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Sequence details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NW_017264853.1:16178-17746'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list sequence details\n",
    "inulinase.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqfromfile = inulinase.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NW_017264853.1:16178-17746\n",
      "Name: NW_017264853.1:16178-17746\n",
      "Description: NW_017264853.1:16178-17746 FOXG_17513 [organism=Fusarium oxysporum f. sp. lycopersici 4287] [GeneID=28958268] [chromosome=Unknown]\n",
      "Number of features: 0\n",
      "Seq('ATGATTACATATCTCAGTCACAAGGCCTGGTTCATGCCTTATCCTACGAAGTTC...TGA')\n"
     ]
    }
   ],
   "source": [
    "# list sequence details\n",
    "for record in SeqIO.parse(\"gene.fna\", \"fasta\"):\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGATTACATATCTCAGTCACAAGGCCTGGTTCATGCCTTATCCTACGAAGTTCGTGGGCCTTGCTTTAATCTCTCTTCTCACTTGCCTCACCCAAGCTGACGACTTTCGTCCTCTTTATCACTTCGTCCCAGAACAAAACTGGATGAATGAACCCAATGGCCTCATCAAGATCGGTTCCCAATGGCACCTCTTCTTCCAGCACAATCCCACCGGAAACTTTTGGGGTAACCTGAGCTGGGGCCATGCGACCAGTACTGACTTGATTGACTGGAACCATCTCCCTGTCGCAATCTCGAGTGCCAATGGGGTTCAAGCCTTCACAGGGACCTCGTACTTTGATGAGGCTAATAGATCTGGTCTTGGAACATCTGAAAACCCACCGTACCTTGCCTTTTATACTGGGTACTTTCCAGACACCGGAGTACAGGATCAGAGACTTGCTTATAGCCTAGACCAAGGGGAGACTTGGATCAAGTACAGCGAAAACCCGATCATCTCGCAGGCACAGGAGAAACCGCACGACATTACCGGTGGCTTAGAGACGCGAGACCCCAAGGTCTTCTTTGACGCTGCAACAAAAACTTGGGTCATGGTCCTCGCCCATGGAGGACAAAACAAATTGTCATTCTGGACATCTCCTGACGCCAAGACTTGGACCTGGCATAACGATCTCACCGATGACGGTATTCCCGGACTTCCCAGTGGTATCACAGGCTGGGAAGTGCCTGATCTCTTTAATCTACCCGTCGAGGGTTCTTCAGACACTAAATGGGTCCTGATAGTTACTCCTGCCCAGGGGTCTCCCGCGGGTGGCAATGGCGTTTTTGCAGTCACTGGCTCGTTTGACGGCGCCACCTTCACCCCCGACCCTGTTGATCCAACCAACATGTGGCTAGATTTTGGGCGTGATTGGGATGGCGCTTATAGCTGGGAAAATGTGCCCACGTCCGATGGCCGTAAGATCCTTGCTTCGGTCATGAACAGCTATGGAGTCGACCCGCCAACCAATGCTTGGAAGGGCATGCTGTCGTTCCCACGTACCTTGCAGCTGCGAGAGATTGGTGGCGAGTTGCGCTTTATTCAGCAGCCTGTCATCGAACTCGAATCTGTCGGTACATCTCTTGGCAACATAGCAAATCAGACTCTCGAGCCAGGGACCACGCTTTTGTCTGACTGTCGCGGAACTGCGTTAGATATCAGCATCTCTTTTGTTCCTACTGAAGGCTCCATCCTATCCCTAGCAGTCCGCAAAGGAGACTCGGAACAAACCGTCATCCAGTATGCTCAATCCAATGGCGCCCTGACTGTTGATCGCCGTGCCAGTGGCAATATCTCATACAACTCGGCGGCCGGTGGTATTCACACAGCAACATTCTCGCCAGATTCCAACAATGTAGCACACCTGCGGGTTCTGGTAGACATGTGTTCGGTGGAGGTATTTGGGGGAGCTGGCGAAGTGGTCATTTCGGACCTGATATTTCCAAGCGAAAGCTCTAACAGACTGTCTCTATCCACGACTGGAGGAAATGTCGTCTTACAGTCCGTTGAAATACGCTCTGTCGCATGA\n"
     ]
    }
   ],
   "source": [
    "# store sequence for later analysis\n",
    "seqfromfile = record.seq\n",
    "print(seqfromfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1569"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# length of sequence\n",
    "len(seqfromfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "483605.91380000464"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#molecular weight\n",
    "molecular_weight(seqfromfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Sequence manipulation using Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1. Indexing / Slicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGA')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CODON\n",
    "seqfromfile[0:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGTGA')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqfromfile[0:3]+seqfromfile[-3:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2. Concatination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATTCATGA')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat = seqfromfile[0:6]+seqfromfile[-5:]\n",
    "concat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3. Codon Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "131"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqfromfile.find('AGA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4. GC Content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In molecular biology and genetics, GC-content is the percentage of nitrogenous bases in a DNA or RNA molecule that are either guanine or cytosine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51.625239005736134"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the logic of GC content\n",
    "(seqfromfile.count('G')+seqfromfile.count('C'))/len(seqfromfile)*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51.62523900573614"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# determineation of GC content using Biopython\n",
    "GC(seqfromfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5. Complement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- The four bases in DNA are adenine (A), cytosine (C), guanine (G), and thymine (T). \n",
    "- These bases form specific pairs (A with T, and G with C). \n",
    "- Base pair may also refer to the actual number of base pairs, such as 8 base pairs, in a sequence of nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATTACATA')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqfromfile[:11]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TACTAATGTAT')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#complement\n",
    "seqfromfile[:11].complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TATGTAATCAT')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#reverse complement\n",
    "seqfromfile[:11].reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TCATGCGACA')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqfromfile[-10:].reverse_complement()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Transcription & Translation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- <b>In transcription</b>, the strand of DNA that is used to synthesize mRNA is known as the template strand. \n",
    "- Whereas, the non-template or coding strand matches the sequence of the RNA. \n",
    "- However, it doesn’t match it exactly as RNA has uracil instead of thymine.\n",
    "- The nucleotides of RNA are known as ribonucleotides. \n",
    "- These nucleotides bond to the template strand via hydrogen bonds after the DNA molecule opens up. \n",
    "- Those nucleotides are bonded together with a phosphodiester bond just like DNA is bonded.\n",
    "\n",
    "\n",
    "- While transcription is the process of creating mRNA from DNA, <b>translation</b> is the process of converting the genetic information of mRNA into proteins.\n",
    "\n",
    "<a align='center'>\n",
    "    <img src='https://images.squarespace-cdn.com/content/v1/52668d02e4b0f593739ec2b6/1533131113081-4Q921GDMF0MW07F1P9OS/Slide01.png?format=750w' width='600'>\n",
    "</a>\n",
    "<a align='center'>\n",
    "    <img src='https://www.nature.com/scitable/content/ne0000/ne0000/ne0000/ne0000/105292500/44376_38a.jpg' width='600'>\n",
    "</a>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1. Transcription"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AUGAUUACAU')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Transcribe\n",
    "mRNA = seqfromfile.transcribe()\n",
    "mRNA[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('ATGATTACAT')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Back transcribe\n",
    "mRNA[:10].back_transcribe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2. Translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(CodonTable.unambiguous_dna_by_id[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MITYLSHKAW')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_seq = seqfromfile.translate()\n",
    "protein_seq[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "523"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(protein_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1569"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqfromfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqfromfile)/len(protein_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQNWMNEPNG...VA*')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_seq.back_transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(CodonTable.unambiguous_dna_by_id[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('G', 50),\n",
       " ('S', 48),\n",
       " ('L', 45),\n",
       " ('T', 43),\n",
       " ('V', 35),\n",
       " ('A', 34),\n",
       " ('P', 32),\n",
       " ('D', 32),\n",
       " ('N', 28),\n",
       " ('I', 27)]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Listing the most common amino acids\n",
    "common_amino = Counter(protein_seq)\n",
    "common_amino.most_common(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Frequency')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAbDElEQVR4nO3de7hdVXnv8e+PICQBKyA7IdyyBVIQEAIEUdDDJXCKCBJowaTUxkpJPQdUtB6NaGu00kM5B5FToBgRCFghKASijzdIHwQttwDhkoJGIUAgkk0Ak3CLJO/5Y46FMzv7MtfOnmtfxu/zPOtZa17GHO+ae693jTXmnGMqIjAzs3xsNtABmJlZaznxm5llxonfzCwzTvxmZplx4jczy4wTv5lZZpz4rWUkLZZ0xEDHYZY7J/7MSFoq6VVJayQ9J+lKSVv3cVshaY+q60fEPhFxWx/rOlHSIkmrJD0vaYGk9r5sayiR1J7285rS48GBjsuGNif+PJ0QEVsDBwIHA1/qvIKkzVseVTfSl8vVwN8DbwPeAVwKrB/IuFpsm4jYOj3277xwMP29bPBz4s9YRDwD/BjYF95swZ8paQmwJM07Q9JvJL0gab6kHdP829NmHkyt0A+n+cenlvlLkv5T0n6N+tKvjaPT61mSrpd0taTVqRtoUjehTgSeiIgFUVgdETdExFNpW5tJminpt5JWpu1uV6r3I5KeTMu+2CmOqyR9rbTuEZKWlaZ3lHSDpA5JT0j6ZGlZj+9B0i6SbkxlV0q6uLTsY5IelfSipJ9KGt/En+7NOCV9XtLvgCuH0n6QtGX6n3pXab0x6ddoWzP7wprnxJ8xSbsAxwEPlGZPAQ4B9pZ0FPC/gVOBccCTwHUAEfHf0vr7p1boXEkHAlcAfwe8HfgmMF/Slt2E8KG0vW2A+cDF3ax3P7CXpAslHamNu6Y+meI+HNgReBG4JL3HvYF/Az6Slr0d2LnbnVIiaTPgB8CDwE7AZOBsSX/W23uQNAL4IcU+a0/lr0vLpgDnACcDbcAdwLVVYupkB2A7YDwwgyG0HyLi9bT+X5W2MQ24NSI6qu4A66OI8COjB7AUWAO8RPFhvBQYlZYFcFRp3W8D55emtwb+ALSX1t+jtPzfgH/qVN+vgMNLdR+dXs+i+JA31tsbeLWHuN8DXA90AK8BVwFbp2WPApNL645LcW4O/CNFomks2wpYW4rjKuBrpeVHAMvS60OApzrF8QXgyt7eA/DeFOvmXbyXHwOnl6Y3A14Bxnexbnvazy+VHp9Nca4FRpbWHWr74RDgaWCzNL0QOHWgPyM5PNwvmKcpEXFrN8ueLr3ekaK1DUBErJG0kqLVtrSLsuOB6ZI+UZq3RdpOV35Xev0KMFLS5hHxRucVI+Iuil8eSDoYmAt8kSIBjQfmSSr3+a8Dxqa6ny5t5+X0HqoYD+wo6aXSvBEULfQe3wOwC/BkV+8lbfciSReU5olivz7ZTSzbl7el4uyojoh4rdN2h8x+iIi7Jb0MHC5pObAHxa8Fq5kTv3VWHq71WYoPPQCStqLoInimm7JPA+dGxLn1hQcRca+kG0nHJlK9H4uIX3ZeNyWUd5amR1O8h4aXgdGl6R1Kr5+mOLYwoQ9hPg3s2s0XWWM//XsftlvWeWjdobYfAOZQdPf8Dvh+py8yq4n7+K0n3wX+RtLE1E//z8DdEbE0LX8O2K20/reAj0s6RIWtJH1Q0ls3JQhJ71NxkHlMmt6Lok/5rrTKZcC5jQOkktoknZiWfR84Pm1jC+CrbPh/vwg4TtJ2knYAzi4tuwdYlQ6gjpI0QtK+6RdHb+4BlgPnpf0wUtJhpXi/IGmfFO/bJJ3SzD7pxlDbDwDXACdRJP+rm37H1idO/NatiFgA/ANwA8WHd3dgammVWcAcFWfwnBoRC4EzKA7svQj8BvhoP4TyEkWif1jSGuAnwDzg/LT8Ioougp9JWk3xhXBIeg+LgTMpvsSWp7iWlbZ9DcVBy6XAzyi6kEhl1wEnkM4qAp4HLqc4pbRHpbJ7AE+lOj+cls0D/gW4TtIq4BHgA1V3Rg+G1H5Iy5dRdCcGG3YdWY2UDqqYZUPSUuBvezjOkYXBsh8kXQE8GxEbXU9i9XAfv5kNGBVXX58MHDDAoWTFXT1mNiAk/RNFN9f/iYgnBjqenLirx8wsM27xm5llZkj08W+//fbR3t4+0GGYmQ0p99133/MRsdHYR0Mi8be3t7Nw4cKBDsPMbEiR1OWV4O7qMTPLjBO/mVlmnPjNzDLjxG9mlhknfjOzzDjxm5llptbTOdMgUKspbgbxRkRMUnEP0LkUdxZaSnHHnRfrjMPMzP6oFS3+IyNiYkQ0br48E1iQbuqwIE2bmVmLDERXz4kUd90hPU8ZgBjMzLJVd+IPiptC3CdpRpo3NiKWA6TnMV0VlDRD0kJJCzs6OmoO02x4kZp7WF7qHrLhsIh4Nt0y7xZJj1UtGBGzgdkAkyZN8hCiZmb9pNYWf0Q8m55XUNwq793Ac5LGAaTnFXXGYGZmG6ot8acbK7+18Rr47xQ3XZgPTE+rTQdurisGMzPbWJ1dPWOBeSo6EDcHvhsRP5F0L3C9pNMpbr58So0xmJlZJ7Ul/oh4HNi/i/krgcl11WtmZj3zlbtmZplx4jczy8yQuAOX1avZ87jDJ9eaDWlu8ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMOPGbmWXGid/MLDNO/GZmmfFYPcNIM2PueLwds3y5xW9mlhknfjOzzDjxm5llxonfzCwzTvxmZplx4jczy4wTv5lZZpz4zcwy4wu4auCbl5vZYOYWv5lZZpz4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8tM7Ylf0ghJD0j6YZreTtItkpak523rjsHMzP6oFS3+TwGPlqZnAgsiYgKwIE2bmVmL1Jr4Je0MfBC4vDT7RGBOej0HmFJnDGZmtqG6W/zfAD4HrC/NGxsRywHS85iuCkqaIWmhpIUdHR01h2lmlo/aEr+k44EVEXFfX8pHxOyImBQRk9ra2vo5OjOzfNU5OudhwIckHQeMBP5E0neA5ySNi4jlksYBK2qMwczMOqmtxR8RX4iInSOiHZgK/EdE/BUwH5ieVpsO3FxXDGZmtrGBOI//POAYSUuAY9K0mZm1SEtuxBIRtwG3pdcrgcmtqNfMzDbmK3fNzDLjxG9mlhknfjOzzDjxm5llxonfzCwzTvxmZplx4jczy4wTv5lZZpz4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8tMS8bjN7O+kaqvG1FfHDa8uMVvZpYZJ34zs8w48ZuZZcZ9/DbkNNPvDe77NuvMLX4zs8w48ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWac+M3MMuPz+M1q5usOeud91FqVWvyS9q07EDMza42qXT2XSbpH0v+UtE2dAZmZWb0qJf6IeB9wGrALsFDSdyUdU2tkZmZWi8oHdyNiCfAl4PPA4cD/k/SYpJPrCs7MzPpf1T7+/SRdCDwKHAWcEBHvTK8v7KbMyNQ99KCkxZK+kuZvJ+kWSUvS87b99F7MzKyCqi3+i4H7gf0j4syIuB8gIp6l+BXQldeBoyJif2AicKyk9wAzgQURMQFYkKbNzKxFqp7OeRzwakSsA5C0GTAyIl6JiGu6KhARAaxJk29JjwBOBI5I8+cAt1F0H5mZWQtUbfHfCowqTY9O83okaYSkRcAK4JaIuBsYGxHLAdLzmKYiNjOzTVI18Y+MiEbrnfR6dG+FImJdREwEdgbe3cz1AJJmSFooaWFHR0fVYmZm1ouqif9lSQc2JiQdBLxatZKIeImiS+dY4DlJ49J2xlH8GuiqzOyImBQRk9ra2qpWZWZmvaia+M8GvifpDkl3AHOBs3oqIKmtcbGXpFHA0cBjwHxgelptOnBz82GbmVlfVTq4GxH3StoL2BMQ8FhE/KGXYuOAOZJGUHzBXB8RP5R0J3C9pNOBp4BT+h6+mZk1q5lB2g4G2lOZAyQREVd3t3JEPAQc0MX8lcDkJuM0M7N+UinxS7oG2B1YBKxLswPoNvGbmdngVLXFPwnYO52bb2ZmQ1jVg7uPADvUGYiZmbVG1Rb/9sB/SbqHYigGACLiQ7VEZWY2iA31G8dUTfyz6gzCzMxap+rpnD+XNB6YEBG3ShoNjKg3NDMzq0PVYZnPAL4PfDPN2gm4qaaYzMysRlUP7p4JHAasgjdvyjLsB1eTmnuYmQ0FVRP/6xGxtjEhaXOK8/jNzGyIqZr4fy7pHGBUutfu94Af1BeWmZnVpWrinwl0AA8Dfwf8iO7vvGVmZoNY1bN61gPfSg8zMxvCqo7V8wRd9OlHxG79HpGZmdWqmbF6GkZSDKW8Xf+HY2ZmdavUxx8RK0uPZyLiG8BR9YZmZmZ1qNrVc2BpcjOKXwBvrSUis0FoqI/NYlZWtavngtLrN4ClwKn9Ho2ZmdWu6lk9R9YdiJmZtUbVrp7P9LQ8Ir7eP+GYmVndmjmr52Bgfpo+AbgdeLqOoMzMrD7N3IjlwIhYDSBpFvC9iPjbugIzM7N6VB2yYVdgbWl6LdDe79GYmVntqrb4rwHukTSP4grek4Cra4vKzMxqU/WsnnMl/Rh4f5r1NxHxQH1hmZlZXap29QCMBlZFxEXAMknvqCkmMzOrUdVbL34Z+DzwhTTrLcB36grKzMzqU7XFfxLwIeBlgIh4Fg/ZYGY2JFVN/GsjIkhDM0vaqr6QzMysTlUT//WSvglsI+kM4FZ8UxYzG+Kk6o/hpNezeiQJmAvsBawC9gT+MSJuqTk2MzOrQa+JPyJC0k0RcRDgZG9mNsRV7eq5S9LBtUZiZmYtUfXK3SOBj0taSnFmjyh+DOxXV2BmZlaPHhO/pF0j4ingA81uWNIuFMM67ACsB2ZHxEWStqM4ZtBOuqFLRLzY7PbNzKxveuvquQkgIp4Evh4RT5YfvZR9A/j7iHgn8B7gTEl7AzOBBRExAViQps3MrEV6S/zlk5h2a2bDEbE8Iu5Pr1cDjwI7AScCc9Jqc4ApzWzXzMw2TW+JP7p53RRJ7cABwN3A2IhYDsWXAzCmmzIzJC2UtLCjo6OvVZtZC+V6XvxQ01vi31/SKkmrgf3S61WSVktaVaUCSVsDNwBnR0SlMgARMTsiJkXEpLa2tqrFzMysFz0e3I2IEZuycUlvoUj6/x4RN6bZz0kaFxHLJY0DVmxKHWZm1pxmhmVuSrri99vAo51uxj4fmJ5eTwdurisGMzPbWNXz+PviMOAjwMOSFqV55wDnUYz9czrwFHBKjTGYmVkntSX+iPgFG54VVDa5rnrNzKxntXX1mJnZ4OTEb2aWGSd+M7PMOPGbmWXGid/MLDNO/GZmmanzPH6zQafZMWKizyNUmQ1ebvGbmWXGid/MLDNO/GZmmXHiNzPLjA/umtmQ5gP2zXOL38wsM078ZmaZceI3M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMOPGbmWXGid/MLDNO/GZmmXHiNzPLjMfqGWRyGnekmfc6lN+n2WDjFr+ZWWac+M3MMuPEb2aWGSd+M7PMOPGbmWXGid/MLDNO/GZmmfF5/LZJcrruwGy4qK3FL+kKSSskPVKat52kWyQtSc/b1lW/mZl1rc6unquAYzvNmwksiIgJwII0bWZmLVRb4o+I24EXOs0+EZiTXs8BptRVv5mZda3VB3fHRsRygPQ8prsVJc2QtFDSwo6OjpYFaGY23A3as3oiYnZETIqISW1tbQMdjpnZsNHqxP+cpHEA6XlFi+s3M8teqxP/fGB6ej0duLnF9ZuZZa/O0zmvBe4E9pS0TNLpwHnAMZKWAMekaTMza6HaLuCKiGndLJpcV51mZta7QXtw18zM6uHEb2aWGSd+M7PMDPtB2jyImJnZhtziNzPLjBO/mVlmnPjNzDLjxG9mlhknfjOzzDjxm5llxonfzCwzTvxmZplx4jczy4wTv5lZZpz4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWac+M3MMjPs77lrZs3xfaqHP7f4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWZ8AZeZWQsNhgvkBqTFL+lYSb+S9BtJMwciBjOzXLU88UsaAVwCfADYG5gmae9Wx2FmlquBaPG/G/hNRDweEWuB64ATByAOM7MsDUQf/07A06XpZcAhnVeSNAOYkSbXSPpVP8exPfD8xvXWWnYg6uyy7EDUWbHsoNlHFcsOmngH8T7qsuwgjnfQ7KMmynZnfFczByLxd/U2Njp8ERGzgdm1BSEtjIhJrSw7EHVuStlc6tyUso633rK51LmpZZs1EF09y4BdStM7A88OQBxmZlkaiMR/LzBB0jskbQFMBeYPQBxmZllqeVdPRLwh6Szgp8AI4IqIWNzqONi0bqS+lh2IOjelbC51bkpZx1tv2Vzq3NSyTVH49jlmZlnxkA1mZplx4jczy0xWiV9SSLqmNL25pA5JP2xiG2uarPPtkhalx+8kPVOa3qKXspL0C0kfKM07VdJP6oi1U9kvSlos6aEU60bXWvRWn6SPSrq42VglHSdpiaRdm4j3pPT33avCuhdKOrs0/VNJl5emL5D0mV62sa70d1wkqb2JWBtlH5H0PUmjK5S5TdKfdZp3tqRLK9Y5VtJ3JT0u6T5Jd0o6qWLZHSRdJ+m3kv5L0o8k/WmFcjtLujn9LR+XdLGkLSuUa+yfxZIelPQZSZVyVad9+wNJ21Qp16ls41FpOJn0f3dBafqzkmY1Ua/S86zydJ2ySvzAy8C+kkal6WOAZ+qsMCJWRsTEiJgIXAZc2JhOVy73VDaAjwNflzRS0lbAucCZdcYs6b3A8cCBEbEfcDQbXnRXZ92TgX8Fjo2Ip5ooOg34BcVZYr35T+DQVN9mFBfO7FNafijwy1628Wrp7zgxIpY2EWuj7L7AWoq/cW+uZeP3NjXN71FKJDcBt0fEbhFxUCq7c8Wy84DbImL3iNgbOAcYW6HcjcBNETEBmACMAs7vrU7+uH/2ofiMHgd8uUK5ctl9gRdo7rPS+W96XsVyrwMnS9q+ibrKTpP0OWBkej6tj9upLLfED/Bj4IPp9TQqfHAGUkQ8AvwA+DzFP//VEfHbmqsdBzwfEa+nGJ6PiNqvtZD0fuBbwAebeY+StgYOA06nWuL/JSnxUyT8R4DVkrZNLdJ3Ag80E/smuAPYo8J63weOb7SY0y+MHSm+7HpzFLA2Ii5rzIiIJyPiXyuUPRL4Q6eyiyLijgp1vhYRV6Yy64BPA3+d/l6VRMQKiiv4z+pDS/hOipEC6vYGxRk5n+5L4Yj4DkXD6nPAU2m6Vjkm/uuAqZJGAvsBdw9wPFV8BfhLioHtqrSYNtXPgF0k/VrSpZIOr1huVPmnMvDVJurcErgZmBIRjzUZ7xTgJxHxa+AFSQf2tHL6EnsjdSUdSpEg7gbeC0wCHurt1xgbvtd5TcYLFF2NFH/Th3tbNyJWAvcAx6ZZU4G5Ue20vH2A+/sSI7AvcF8fyu3TuVxErAKWUu2LrlzucYpcNaZqGRWDQU6muWuENvj/lfThJspeQtFyf1sTZQCQ9JcUF7WeD+yapmuV3Xj8EfFQai1NA340wOFUEhEvS5oLrGm0wmuub42kg4D3U7T45kqaGRFX9VL01dSlBRR9/BSJtIo/UHTBnA58qsmQpwHfSK+vS9O9JbpGq/9Q4OsULcNDgd+nOHqzwXtt0qj0xQhFi//bFcs1untuTs8f60vlki4B3kfxK+DgvmyjSjV0MRQLXQ/ZUnV7VTT2bTvFF88tTdTR579pRKySdDXwSeDVJotfGxEhaVZEnO8+/vrMB/4vg7ybp5P16dESEbEuIm6LiC8DZwF/XnOV64FTgYMlnVO1kKS3U3QrXC5pKfC/gA9X+PA0+vnfRdHVcxdFi79K//6mKvclf6LCr4uGm4DJ6RfNqIio2opfDLz5KygizqRoDbdVLHtQxXo6l9vgS1/Sn1AcG2hqwEVJuwHrgBUVVm8k7/HAFtR8PKyTb1A0XLZqplDjV1tEzCpP1ynXxH8F8NWI6PUndo4k7SlpQmnWRODJuuuNiFcoDiqfJun0isX+guK4x/iIaI+IXYAnKFq0PfllquuF9CX3ArANRfK/s09voGYRsQa4jeL/t5lGy39QHDj8H6V5vZ5JVCq7paQzGjMkHVyh+28BMFrSX6cyI4ALgIsjonKLWFIbxUkRFzeTECPi9xSt789KekvVcpsi/Q9dT5H8B7UsE39ELIuIiwY6jhYYLWlZ6dHjKYolWwNz0ql7D1HcMGdWbVGWpA/PscCXJFW5T8M0irNOym6gOCbSk4cpzua5q9O830fERkPjDiLXAvtTdGlVkhLmFOBwSU9IugeYQ3HCQJWyJwHHpNM5F1P8L/R4sL9U7i8kLQFWAusj4twKITf62hcDt1Icc/pKhXKdY3gAeJBqB/zL9TYeVc/qKbuA4v9qUPOQDWZWO0mHUnxpnRwRfTlYbP3Iid/MLDNZdvWYmeXMid/MLDNO/GZmmXHiNzPLjBO/DUtqYrTOLsp+vHH+eT/F8mlJr1W5nF/FyJfbdDF/lqTP9ldMljcnfhuumhmtcwMRcVlEXN3PsdxLcV57b3UfFxEv9WPdZhtx4rdhp7vROiUdIennkq5PA9CdJ+k0SfdIeljS7mm9N1vXKsbB/5e0zq/TCKKkYbKvTOUekHRkN7HsTnFB3JcovgDejLFU/iFJf57mL1Ua3lfFPRF+JelWYM8adpVlKrtB2iwLU0ijdUp6QdKBpXFt9qcYdvkF4HHg8oh4t6RPAZ8Azu5ie5undRrjwh9NGgMmIt6VupN+JulPI+K1TmUbQ3/fAewpaUwaavgfKK4SfheApG3LhdIgeVOBAyg+p/fTt1EyzTbiFr8NR9P445AGjdE6G+6NiOVplNPfUgwHAMVwDe3dbO/G9HxfaZ33AdcApGGknwS6uivVVOC6iFiftnNKmn80xVC+pG282Knc+4F5EfFKGs64meGFzXrkFr8NK6XROveVFMAIIFTc2QiKuyU1rC9Nr6f7z0NjnXWldXodOlfSfhR3nrolDRa6BcWvjEvoftjiMl9Wb7Vwi9+Gm76O1tms20m3yFNx/9ld2Xi44WnArBRHe0TsCOwkaTzFL42zGit27upJ2z9J0ihJbwVO6Of4LWNO/Dbc9HW0zmZdCoyQ9DAwF/hoFzfJmdpFLPPS/K8B26q4KfiDFDe8eVM6JjEXWJTi7+1Wh2aVeZA2M7PMuMVvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWb+P0tm+ntQQGhvAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# visualize all 20 amino acid occurrences in the form of a histogram\n",
    "pylab.bar(common_amino.keys(), common_amino.values(), color='blue')\n",
    "\n",
    "pylab.title('Protein Sequence Frequency')\n",
    "pylab.xlabel('Amino Acid')\n",
    "pylab.ylabel('Frequency')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " - Note: Because stop codon * signifies end of a protein we can split the sequence using (*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQNWMNEPNGLIKIGSQWHLFFQHNPTGNFWGNLSWGHATSTDLIDWNHLPVAISSANGVQAFTGTSYFDEANRSGLGTSENPPYLAFYTGYFPDTGVQDQRLAYSLDQGETWIKYSENPIISQAQEKPHDITGGLETRDPKVFFDAATKTWVMVLAHGGQNKLSFWTSPDAKTWTWHNDLTDDGIPGLPSGITGWEVPDLFNLPVEGSSDTKWVLIVTPAQGSPAGGNGVFAVTGSFDGATFTPDPVDPTNMWLDFGRDWDGAYSWENVPTSDGRKILASVMNSYGVDPPTNAWKGMLSFPRTLQLREIGGELRFIQQPVIELESVGTSLGNIANQTLEPGTTLLSDCRGTALDISISFVPTEGSILSLAVRKGDSEQTVIQYAQSNGALTVDRRASGNISYNSAAGGIHTATFSPDSNNVAHLRVLVDMCSVEVFGGAGEVVISDLIFPSESSNRLSLSTTGGNVVLQSVEIRSVA',\n",
       " '']"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_list = [str(i) for i in protein_seq.split('*')]\n",
    "protein_list[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQ...</td>\n",
       "      <td>522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         protein_seq  length\n",
       "0  MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQ...     522"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert sequences to dataframe\n",
    "large_proteins = [x for x in protein_list if len(x)>2]\n",
    "df = pd.DataFrame({'protein_seq':large_proteins})\n",
    "\n",
    "# Add a column with sequence lengths\n",
    "df['length'] = df['protein_seq'].apply(len)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQ...</td>\n",
       "      <td>522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         protein_seq  length\n",
       "0  MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQ...     522"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sort sequence data\n",
    "df.sort_values(by = ['length'], ascending=False)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQNWMNEPNGLIKIGSQWHLFFQHNPTGNFWGNLSWGHATSTDLIDWNHLPVAISSANGVQAFTGTSYFDEANRSGLGTSENPPYLAFYTGYFPDTGVQDQRLAYSLDQGETWIKYSENPIISQAQEKPHDITGGLETRDPKVFFDAATKTWVMVLAHGGQNKLSFWTSPDAKTWTWHNDLTDDGIPGLPSGITGWEVPDLFNLPVEGSSDTKWVLIVTPAQGSPAGGNGVFAVTGSFDGATFTPDPVDPTNMWLDFGRDWDGAYSWENVPTSDGRKILASVMNSYGVDPPTNAWKGMLSFPRTLQLREIGGELRFIQQPVIELESVGTSLGNIANQTLEPGTTLLSDCRGTALDISISFVPTEGSILSLAVRKGDSEQTVIQYAQSNGALTVDRRASGNISYNSAAGGIHTATFSPDSNNVAHLRVLVDMCSVEVFGGAGEVVISDLIFPSESSNRLSLSTTGGNVVLQSVEIRSVA'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's take a single protein from the table\n",
    "one_large_protein = df.nlargest(1, 'length')\n",
    "gene_inulinase = one_large_protein.iloc[0,0]\n",
    "gene_inulinase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write to a file\n",
    "with open('gene_inulinase.fasta', 'w') as file:\n",
    "    file.write('>large protein\\n'+gene_inulinase)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Basic Local Alignment Using NCBI-BLAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MITYLSHKAWFMPYPTKFVGLALISLLTCLTQADDFRPLYHFVPEQNWMNEPNG...SVA')"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read single_seq.fasta\n",
    "read = SeqIO.read('gene_inulinase.fasta','fasta')\n",
    "read.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 1min 4s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# based on the server load this query might take 2-3 minutes to run\n",
    "result_handle = NCBIWWW.qblast('blastp','pdb',read.seq)\n",
    "blast_qresult = SearchIO.read(result_handle,'blast-xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program: blastp (2.13.0+)\n",
      "  Query: unnamed (522)\n",
      "         protein product\n",
      " Target: pdb\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  pdb|3RWK|X  First crystal structure of an endo-inulinas...\n",
      "            1      1  pdb|1Y4W|A  Crystal structure of exo-inulinase from Asp...\n",
      "            2      1  pdb|3U14|A  Chain A, Fructofuranosidase [Schwanniomyces...\n",
      "            3      1  pdb|3KF3|A  Chain A, Invertase [Schwanniomyces occident...\n",
      "            4      1  pdb|3KF5|A  Chain A, Invertase [Schwanniomyces occident...\n"
     ]
    }
   ],
   "source": [
    "print(blast_qresult[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence ID:pdb|3RWK|X\n",
      "description:First crystal structure of an endo-inulinase, from Aspergillus ficuum: structural analysis and comparison with other GH32 enzymes. [Aspergillus ficuum]\n",
      "E value:    0.0 \n",
      "Bit Score:  696.812\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#fetch the id, description, evalue, bitscore & alignment of first hit\n",
    "\n",
    "seqid = blast_qresult[0]\n",
    "\n",
    "details = seqid[0]\n",
    "\n",
    "print(f\"\\\n",
    "Sequence ID:{seqid.id}\\n\\\n",
    "description:{seqid.description}\\n\\\n",
    "E value:    {details.evalue} \\n\\\n",
    "Bit Score:  {details.bitscore}\\n\\\n",
    "\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alignment:\n",
      "Alignment with 2 rows and 509 columns\n",
      "LALISLLTCL-------TQADDFRPLYHFVPEQNWMNEPNGLIK...SVA unnamed\n",
      "VAYMVWMTCLGLTLPSQAQSNDYRPSYHFTPDQYWMNEPNGLIK...SVS pdb|3RWK|X\n"
     ]
    }
   ],
   "source": [
    "print(f\"alignment:\\n{details.aln}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Reading PDB file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retreiving PDB Structure From RCSB PDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'pdb|3RWK|X'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqid.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3RWK'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# split seqid\n",
    "seqid.id.split('|')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('C:/Users/test/Desktop/PROGRAMMING_WEB DEVELOPMENT/IBM_DATA_SCIENCE_PYTHON/11. BIO_PYTHON/3RWK_inulinase.pdb',\n",
       " <http.client.HTTPMessage at 0x20d04862550>)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# link format https://files.rcsb.org/download/3RWK.pdb\n",
    "urllib.request.urlretrieve('https://files.rcsb.org/download/3RWK.pdb','C:/Users/test/Desktop/PROGRAMMING_WEB DEVELOPMENT/IBM_DATA_SCIENCE_PYTHON/11. BIO_PYTHON/3RWK_inulinase.pdb' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.1 Reading PDB file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\test\\anaconda3\\lib\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain X is discontinuous at line 4361.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Structure id=3RWK>"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parser = PDBParser()\n",
    "structure = parser.get_structure('3RWK', '3RWK_inulinase.pdb')\n",
    "structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.2 Identifying the number of chains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chain ID: X\n",
      "chain ID: A\n",
      "chain ID: B\n"
     ]
    }
   ],
   "source": [
    "for chain in structure[0]:\n",
    "    print(f'chain ID: {chain.id}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. <i>Fusarium oxysporum</i> Inulinase Protein structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1. nglview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0f43316262b241a4b8ed1c7c8399c854",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nv.demo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "739c205737a44527978105b564a5498f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view = nv.show_biopython(structure)\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "17188e99c8584974abd593cf1382f9fb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Image(value=b'', width='99%')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view.render_image()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.2. nglview GUI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "930b2e603ffc4a08823a121f3b91deaa",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8b1601a1c70d4d4190021d600b2a05ce",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Tab(children=(Box(children=(Box(children=(Box(children=(Label(value='step'), IntSlider(value=1, min=-100)), la…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#GUI\n",
    "nv.show_biopython(structure, gui=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions<br>\n",
    "<div style=\"Font-size:16px\">\n",
    "- The inulinase sequence of <i>Fusarium oxysporum</i> has 1569 base pairs<br>\n",
    "- The GC content of the sequence is 51.63<br>\n",
    "- Largest protein BLAST results corresponds to <b>inulinase</b>3RWK <br>\n",
    "- Protein 3RWK has 3 chains inclusing, X, A, and B<br>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "05811a4c8edeb62e87c3e708461cae05c65dc4bb514779bc955273fc84b8bc5c"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
